manual/src/tutorials/effectexamples.etex

```diff
@@ -0,0 +1,666 @@
+segment pointed to by "k" to the current stack. Since neither continuation
+capture nor resumption requires copying stack frames, suspending the execution
+using "perform" and resuming it using either "continue" or "discontinue" are
+fast.
+
+\subsection{s:effects-unhandled}{Unhandled effects}
+
+Unlike languages such as Eff and Koka, effect handlers in OCaml do not provide
+\textit{effect safety}; the compiler does not statically ensure that all the
+effects performed by the program is handled. If effects do not have a matching
+handler, then "Unhandled" exception is raised at the point of the corresponding
+"perform". For example, in the previous example, "bar" does not handle the
+effect "F". Hence, we will get "Unhandled" exception when we run "bar".
+
+\begin{caml_example}{verbatim}
+try bar () with Unhandled -> "Saw unhandled exception"
+\end{caml_example}
+
+\subsection{s:effects-linearity}{Linear continuations}
+
+The delimited continuations in OCaml must be used linearly -- every captured
+continuation must be resumed either with a "continue" or "discontinue" exactly
+once. Attempting to use a continuation more than once raises
+"Continuation_already_taken" exception. For example:
+
+\begin{caml_example}{verbatim}
+try_with perform (Xchg 0)
+{ effc = fun (type a) (eff : a t) ->
+    match eff with
+    | Xchg n -> Some (fun (k: (a, _) continuation) ->
+        continue k 21 + continue k 21)
+    | _ -> None }
+\end{caml_example}
+
+The primary motivation for adding effect handlers to OCaml is to enable
+concurrent programming. One-shot continuations are sufficient for almost all of
+the concurrent programming use cases. They are also much cheaper to implement
+compared to multi-shot continuations since they do not require stack frames to
+be copied. Moreover, OCaml programs may also manipulate linear resources such
+as sockets and file descriptors. The linearity discipline is easily broken if
+the continuations are allowed to resume more than once. It would be quite hard
+to debug such linearity violations on resources due to the lack of static
+checks for linearity and the non-local nature of control flow. Hence, OCaml
+does not support multi-shot continuations.
+
+While the ``at-most once resumption'' property on continuations is ensured with
+a dynamic check, there is no check to ensure that the continuations are resumed
+``at least once''. It is left to the user to ensure that the captured
+continuations are resumed at least once. Not resuming continuations will leak
+the memory allocated for the fibers and also any resources that the suspended
+computation may hold.
+
+One may install a finaliser on the captured continuation to ensure that the
+resources are freed:
+
+\begin{caml}
+\begin{camlinput}
+exception Unwind
+Gc.finalise (fun k ->
+  try ignore (discontinue k Unwind) with _ -> ()) k
+\end{camlinput}
+\end{caml}
+
+In this case, if "k" is never resumed, then the finaliser ensures that the
```

Comment:
Technically, "if `k` is never resumed" should be replaced with "if `k` becomes unreachable".